src/main/java/org/parchmentmc/compass/data/validation/DataValidator.java

```diff
@@ -0,0 +1,222 @@
+            if (!classResult.getData().isEmpty() || !classResult.isEmpty()) {
+                results.addClass(classResult);
+            }
+
+            // Finished visiting one class
+        }
+
+        ctx.postVisit(ctx.classValidators, DataType.CLASSES);
+        // All classes have been visited
+
+        return results;
+    }
+
+    private static class Context {
+        final List<ValidationIssue> packageIssues = new ArrayList<>();
+        final Set<Validator> packageValidators;
+        final List<ValidationIssue> classIssues = new ArrayList<>();
+        final Set<Validator> classValidators;
+        final Set<Validator> currentClassValidators;
+        final List<ValidationIssue> fieldIssues = new ArrayList<>();
+        final Set<Validator> fieldValidators;
+        final List<ValidationIssue> methodIssues = new ArrayList<>();
+        final Set<Validator> methodValidators;
+        final Set<Validator> currentMethodValidators;
+        final List<ValidationIssue> paramIssues = new ArrayList<>();
+        final Set<Validator> paramValidators;
+
+        public Context(int validators) {
+            this.packageValidators = new LinkedHashSet<>(validators);
+            this.classValidators = new LinkedHashSet<>(validators);
+            this.currentClassValidators = new LinkedHashSet<>(validators);
+            this.fieldValidators = new LinkedHashSet<>(validators);
+            this.methodValidators = new LinkedHashSet<>(validators);
+            this.currentMethodValidators = new LinkedHashSet<>(validators);
+            this.paramValidators = new LinkedHashSet<>(validators);
+        }
+
+        public void matching(Set<Validator> workingSet, List<ValidationIssue> issues,
+                             Consumer<Validator> validatorConsumer, Consumer<List<? extends ValidationIssue>> issuesConsumer) {
+            issues.clear();
+            workingSet.forEach(v -> {
+                v.issueHandler = issues::add;
+                validatorConsumer.accept(v);
+                v.issueHandler = null;
+            });
+            if (!issues.isEmpty()) {
+                issuesConsumer.accept(copyOf(issues));
+                issues.clear();
+            }
+        }
+
+        public <C> C removeMatching(Set<Validator> workingSet, Set<Validator> superset, List<ValidationIssue> issues,
+                                    Predicate<Validator> validatorPredicate,
+                                    Function<List<? extends ValidationIssue>, C> resultCreator) {
+            issues.clear();
+            workingSet.clear();
+            workingSet.addAll(superset);
+            workingSet.removeIf(v -> {
+                v.issueHandler = issues::add;
+                boolean walkChildren = validatorPredicate.test(v);
+                v.issueHandler = null;
+                return walkChildren;
+            });
+            C result = resultCreator.apply(copyOf(issues));
```

Comment:
Future work: we currently eagerly create the result object (via the provided function) for classes and methods, because they are used to contain the results from their children (parameters for methods, and methods/fields for classes). The result object contains `TreeMap`s for holding their children, so if the result object is unused, then we've created these `TreeMaps` for nothing. 

We can potentially save on this by creating a reusable lazy holder for these objects, where it only creates the result container if it actually needs to add a child's result.